This project is from the Codecademy Machine Learning career path.

## 18.3.5	A/B Testing at Nosh Mish Mosh  
The Nosh Mish Mosh is a recipe and ingredient meal delivery service. They ship the raw materials and you get to cook them at your home! They’ve decided to hire a data analyst to help make product and interface decisions. Get started to help them figure out the amount of data they’ll need to make meaningful decisions.  
Nosh Mish Mosh: An Assortment of Edible Aliments  
1. We’ve collected customer data for the past week and exposed it through a Python library, so first import `noshmishmosh`.  
2. Next, we’ll need to do a little bit of data analysis — let’s use numpy to help. Import numpy into your workspace as np.
A/B Testing at Nosh Mish Mosh  
3. Nosh Mish Mosh wants to run an experiment to see if we can convince more people to purchase meal plans if we use a more artisanal-looking vegetable selection. We’ve photographed these modern meals with blush tomatoes and graffiti eggplants, but aren’t sure if this strategy will sell enough units to benefit from establishing a business relationship with a new provider.
Before running this experiment, of course, we need to know the sample size that will be required to detect the difference we are hoping for. There are three things we need to know before we can determine that number.  
o	the Baseline Conversion Rate  
o	Minimum Detectable Effect (desired lift)  
o	and the Statistical Significance Threshold  
4. Let’s get the ball rolling on finding those numbers! In order to get our baseline, we need to first know how many users visit the site in a typical week. Let’s grab that logged information, which is stored in `noshmishmosh.customer_visits`. Assign that to a new variable called `all_visitors`.  
5. Next we need to know how many visitors to the site ultimately end up buying a meal or set of meals in a typical week. We have that information saved into `purchasing_customers` field on `noshmishmosh`. Save that information into a variable called `paying_visitors`.  
6. Calculate the lengths of the two lists, saving the results into variables called `total_visitor_count` and `paying_visitor_count`, respectively.  
7. Now to get the baseline: Divide the number of purchasing visitors by the number of total visitors. Save the result in a variable called `baseline_percent`. Since we want a percentage as our answer, multiply the result by 100.0.  
8. Print out the baseline_percent so we know what to use for our baseline percentage in the A/B Sample Size Calculator.  
### Mish Mosh B'Gosh: The Effect Size  
9. These rainbow fingerling potatoes don’t come cheap. We’d like to know for sure that, with this change, we’ll be pulling in at least $1240 more every week. In order to figure out how many more customers we need, we’ll have to investigate the average revenue generated from a given sale. Luckily we have a list of the money spent by each customer in a typical week: `noshmishmosh.money_spent`. Save that list into a variable called `payment_history`.  
10. We need to find how many purchases it would take to reach $1240 in additional revenue using our historical data.
Let’s start with computing the average payment per paying customer using np.mean, saving it as `average_payment`.  
11. We want to know how many of these “usual” payments it would take to clear our $1240 mark. Round the number up using np.ceil (because that’s how many new customers it takes to bring in more than $1240). Save that value into a `new_customers_needed` variable.  
12. Now find the additional percent of weekly visitors who must make a purchase in order to make this change worthwhile. Do this by dividing the number of customers by the total visitor count for a typical week (calculated earlier), and multiplying by 100. Save the result in a variable called `percentage_point_increase`. Print percentage_point_increase to see what it is.  
13. In order to find our **minimum detectable effect/desired lift**, we need to express `percentage_point_increase` as a percent of `baseline_percent`. You can do this by dividing percentage_point_increase by baseline_percent and multiplying by 100.0.
Store the results in a variable called `mde`.  
14. Print out the result mde.  
### Nosh Mish Mosh: Tying It All Together  
15. The last thing we need to calculate the sample size for Nosh Mish Mosh’s artisanal rebranding is our statistical significance threshold. We’d like to be fairly certain, but this isn’t going to be a million dollar decision, so let’s go with 10%.  
16. Now put it all together! Punch the baseline, the minimum detectable effect, and the statistical significance threshold into the calculator and evaluate how many people need to be shown the new assets before we can check if the results are a significant improvement. Save the results in a variable called `ab_sample_size`.


In [1]:
# Step 1
import noshmishmosh

# Step 2
import numpy as np

# Step 4
all_visitors = noshmishmosh.customer_visits

# Step 5
paying_visitors = noshmishmosh.purchasing_customers

# Step 6
total_visitor_count = len(all_visitors)
paying_visitor_count = len(paying_visitors)

# Step 7: Percent of customers.
baseline_percent = paying_visitor_count / total_visitor_count * 100

# Step 8
print(f"Baseline Conversion Rate: {baseline_percent:.2f}%")

# Step 9
payment_history = noshmishmosh.money_spent

# Step 10
average_payment = np.mean(payment_history)

# Step 11
new_customers_needed = np.ceil(1240 / average_payment)



Baseline Conversion Rate: 18.60%


Step 12:  
The additional customers come from the unchanged pool of total visitors. 
Hence, this is the number of visitors that need to be converted to paying visitors or customers.

In [2]:
# Step 12  
percentage_point_increase = new_customers_needed / total_visitor_count * 100
print(f"Percentage Point Increase: {percentage_point_increase:.2f}%")

# Step 13
mde = percentage_point_increase / baseline_percent * 100

# Step 14
print(f"Minimum Detectable Effect (MDE): {mde:.2f}%")

Percentage Point Increase: 9.40%
Minimum Detectable Effect (MDE): 50.54%


Step 16:   
Using the provided baseline calculator with following variables 
* Baseline conversion rate: 18.60%
* Significance threshold: 9.40%
* Minimum detectable effect (MDE): 50.54%

![Alt text](Images/Baseline-Calculator-1.jpg)

In [3]:
# Step 16 
ab_sample_size =  490
print(f"Required Sample Size: {ab_sample_size}")

Required Sample Size: 490
